### ISSUE TITLE
Don't copy styles only to change margins, borders and paddings

### ISSUE THREAD

**Issue by liZe** [Kozea/WeasyPrint#497]
Many situations force to override (often remove) margins, borders and paddings, including:
- table cells with collapsing borders,
- blocks split between pages,
- inline boxes split by block boxes.

Depending on the situation, the style is copied and modified, or the computed values are applied to the box (`box.padding_right` and its friends). There's no perfect solution, as the first one greatly increases memory usage, and the second one breaks `min/max-width` calculation and box copy.

We have to find a solution to:
- never copy nor modify style dicts,
- override margins, borders and paddings on styles without modifying the dict,
- avoid mixing `box.style['margin_left']` and `box.margin_left` (as it's done in `mix/max-width` calculation on boxes after dimensions are already set on the box),
- keep speed (i.e. don't use the copy-on-write method used before, as it was overriding the `__getitem__` method that's over-optimized).

**Comment by liZe** [comment]
In [this wonderful blog entry](https://hacks.mozilla.org/2017/08/inside-a-super-fast-css-engine-quantum-css-aka-stylo/), the solution I was thinking of is given in the chapter called "A sidenote: style struct sharing".

**Comment by liZe** [comment]
4057a92 uses real dicts for performance and use [ChainMap](https://docs.python.org/3/library/collections.html#collections.ChainMap)s to keep original dicts when changing padding/borders/margins, breaking the Python2 compatibility.

**Comment by liZe** [comment]
See https://github.com/Kozea/WeasyPrint/pull/948#issuecomment-534608574 for more information about where to remove style edition. #950 removed the worse parts, but it’s not finished yet.

Having frozen styles with computed values is the way to go.

**Comment by Tontyna** [comment]
> Having frozen styles with computed values is the way to go.

Of course. But grepping for modifications of `Box.style` it looks like since #950 we don't touch the immutable computed values anymore. The modfications rather ensure the correct initialization of computed values for specialized boxes.

What I don't like is the fact that the pagination stores used values (and other properties related to layout) as attributes in the root_box tree and its children. AFAIK it doesn't hurt, not sure whether it's required, but it definitely feels wrong.

A 'style struct sharing' would be great. At the moment each box of the formatted root_box tree has its own complete style dictionary -- style sharing only happens between the formatted box and its paginated counterpart(s).
